File: algorithms/LeetCode/rotate_list.scala
// LeetCode: 61. Rotate List

/**
 * Definition for singly-linked list.
 * class ListNode(var _x: Int = 0) {
 *   var next: ListNode = null
 *   var x: Int = _x
 * }
 */
import scala.util.control.Breaks._

object Solution {
    def rotateRight(head: ListNode, k: Int): ListNode = {
        if(head == null || head.next == null){
            return head;
        }
        var temp_k = k;
        while(temp_k>100000){
            var q = temp_k / 100000;
            var r = temp_k % 100000;
            temp_k = q + r;
        }
        println(temp_k);
        var newHead: ListNode = head;
        var temp: ListNode = newHead;
        var counter = 0;
        var last: ListNode = null;
        var sLast: ListNode = null;
        breakable{
            while(counter <= temp_k){
                while(temp.next != null){
                    sLast = temp;
                    last = temp.next;
                    temp = temp.next;
                }
                

                if(counter == temp_k){
                    break;
                } else {
                    last.next = newHead;
                    newHead = last;
                    temp = newHead;
                    sLast.next = null;
                }
                counter += 1;
            }
        }
        
        return newHead;
    }
}

File: algorithms/LeetCode/rotate_list.py
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def rotateRight(self, head, k):
        """
        :type head: ListNode
        :type k: int
        :rtype: ListNode
        """
        if not head: return None
        l = 1
        p = head
        while p.next:
            p = p.next
            l += 1
        k = k % l
        if l > k: p.next = head
        p = head
        for i in range(l-k-1):
            p = p.next
        res = p.next
        p.next = None
        return res

